src/kaleidoscope/plugin/DynamicTapDance.h

```diff
@@ -0,0 +1,49 @@
+/* DynamicTapDance -- Dynamic TapDance support for Kaleidoscope
+ * Copyright (C) 2019  Keyboard.io, Inc
+ * Copyright (C) 2019  Dygma Lab S.L.
+ *
+ * This program is free software: you can redistribute it and/or modify it under
+ * the terms of the GNU General Public License as published by the Free Software
+ * Foundation, version 3.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <Kaleidoscope.h>
+#include <Kaleidoscope-TapDance.h>
+
+namespace kaleidoscope {
+namespace plugin {
+
+class DynamicTapDance: public kaleidoscope::Plugin {
+ public:
+  DynamicTapDance() {}
+
+  EventHandlerResult onFocusEvent(const char *command);
+
+  static void setup(uint8_t dynamic_offset, uint16_t size);
+
+  static bool dance(uint8_t tap_dance_index, KeyAddr key_addr, uint8_t tap_count,
+                    TapDance::ActionType tap_dance_action);
+
+ private:
+  static uint16_t storage_base_;
```

Comment:
`storage_base_` and `storage_size_`, yes. The first is the offset from the beginning of storage to the start of this plugin's slice. Even on AVR, this can be more than 256 bytes, so won't fit into an `uint8_t`. We can't scale it down either, because then we'd potentially waste precious EEPROM bytes.

Same for `storage_size_`: it's perfectly possible to allocate more than 256 bytes for tap-dance storage. Perhaps not on AVR which has limited EEPROM, but on ARM, where you can have 16k storage, 512 for tapdance is not the end of the world.

FWIW, we use the same sizes for `DynamicMacros`, and `storage_base_` is `uint16_t` for _every_ plugin that uses Storage.

We _could_ make it `uint32_t`, but that'd be wasteful at this point.